imports/plugins/included/taxes-taxcloud/server/hooks/hooks.js

```diff
@@ -1,8 +1,182 @@
+        shopItemsMap[cartItem.shopId] = [];
+      }
+      shopItemsMap[cartItem.shopId].push(taxCloudItem);
+    }
+  }
+  const shippingAddressMap = cartToCalc.shipping.reduce((addressMap, shipping) => {
+    if (!addressMap[shipping.shopId]) {
+      addressMap[shipping.shopId] = shipping.address;
+    }
+    return addressMap;
+  }, {});
+  // Create a map of shopId -> shopAddress
+  const shopsList = Shops.find({ _id: { $in: Object.keys(shopItemsMap) } }, { _id: 1, addressBook: 1 }).fetch();
+  let shopsHaveAddress = true;
+  const shopsAddressMap = shopsList.reduce((addressMap, shop) => {
+    if (!shop.addressBook) {
+      shopsHaveAddress = false;
+      return;
+    }
+    [addressMap[shop._id]] = shop.addressBook;
+    return addressMap;
+  }, {});
+  if (!shopsHaveAddress) {
+    // All the marketplace shops don't have address yet.
+    Logger.error("All the marketplace shops don't have address yet.");
+    return;
+  }
+  let totalTax = 0;
+  const subtotalsByShop = cartToCalc.getSubtotalByShop();
+  const subTotal = cartToCalc.getSubTotal();
+  const responsePromise = [];
+  // For each shop get the tax on the products.
+  Object.keys(shopItemsMap).forEach((shopId) => {
+    const taxCloudSettings = Reaction.getPackageSettingsWithOptions({ name: "taxes-taxcloud", shopId });
+    const { apiKey, apiLoginId } = taxCloudSettings.settings.taxcloud;
+    if (!apiKey || !apiLoginId) {
+      Logger.warn("TaxCloud API Key is required.");
+    }
+    const shopAddress = shopsAddressMap[shopId];
+    const { zip4: originZip4, zip5: originZip5 } = zipSplit(shopAddress.postal);
+    const origin = {
+      Address1: shopAddress.address1,
+      City: shopAddress.city,
+      State: shopAddress.region,
+      Zip5: originZip5,
+      Zip4: originZip4
+    };
+
+    const shippingAddress = shippingAddressMap[shopId];
+    // User hasn't entered shipping address yet.
+    if (!shippingAddress) return;
+    const { zip4: destZip4, zip5: destZip5 } = zipSplit(shippingAddress.postal);
+    const destination = {
+      Address1: shippingAddress.address1,
+      City: shippingAddress.city,
+      State: shippingAddress.region,
+      Zip5: destZip5,
+      Zip4: destZip4
+    };
+
+    // request object
+    const request = {
+      headers: {
+        "accept": "application/json",
+        "content-type": "application/json"
+      },
+      data: {
+        apiKey,
+        apiLoginId,
+        customerID: cartToCalc.userId,
+        cartItems: shopItemsMap[shopId],
+        origin,
+        destination,
+        cartID: cartToCalc._id,
+        deliveredBySeller: false
+      }
+    };
+    // pushing all the requests to array.
+    responsePromise.push(new Promise((resolve, reject) => {
+      HTTP.post(url, request, (error, response) => {
+        // ResponseType 3 is a successful call.
+        if (response.data.ResponseType !== 3) {
+          let errMsg = "Unable to access service. Check credentials.";
```

Comment:
Are we sure that this is a credentials-based error?

Reply:
Yes.